### PR TITLE
Improve loose comparison on integer

### PR DESCRIPTION
**Pull request by staabm**
see https://www.php.net/manual/en/types.comparisons.php

grafik

see https://3v4l.org/395hr

### CODE REVIEW

File: tests/PHPStan/Analyser/nsrt/equal.php [hunk 1]
```diff
@@ -152,7 +152,6 @@ public function doFoo(string $a, int $b, float $c): void
 		assertType('false', 'a' != 'a');
 		assertType('true', 'a' != 'b');
 
-		assertType('bool', $b == 'a');
```
Comment (staabm) [inline tests/PHPStan/Analyser/nsrt/equal.php:155]: moved this case into the loose-comparison-* files, as it depends on php version

File: src/Type/Traits/ConstantScalarTypeTrait.php [hunk 2]
```diff
@@ -58,7 +58,7 @@ public function looseCompare(Type $type, PhpVersion $phpVersion): BooleanType
 		}
 
 		if ($type->isConstantArray()->yes() && $type->isIterableAtLeastOnce()->no()) {
-			// @phpstan-ignore equal.notAllowed, equal.invalid
+			// @phpstan-ignore equal.notAllowed, equal.invalid, equal.alwaysFalse
```
Comment (staabm) [inline src/Type/Traits/ConstantScalarTypeTrait.php:61]: ignores

```
------ ------------------------------------------------------------------------------------------------------------- 
  Line   src/Type/Traits/ConstantScalarTypeTrait.php (in context of class PHPStan\Type\Constant\ConstantIntegerType)  
 ------ ------------------------------------------------------------------------------------------------------------- 
  62     Loose comparison using == between int and array{} will always evaluate to false.                             
         🪪  equal.alwaysFalse                                                                                        
 ------ ------------------------------------------------------------------------------------------------------------- 

```

File: tests/PHPStan/Rules/Comparison/ConstantLooseComparisonRuleTest.php [hunk 3]
```diff
@@ -173,6 +174,24 @@ public function testBug11694(): void
...
+					29,
+					'Because the type is coming from a PHPDoc, you can turn off this check by setting <fg=cyan>treatPhpDocTypesAsCertain: false</> in your <fg=cyan>%configurationFile%</>.',
+				],
+				[
+					"Loose comparison using == between int<10, 20> and '13foo' will always evaluate to false.",
+					30,
+					'Because the type is coming from a PHPDoc, you can turn off this check by setting <fg=cyan>treatPhpDocTypesAsCertain: false</> in your <fg=cyan>%configurationFile%</>.',
+				],
+			]);
+		}
```
Comment (staabm) [inline tests/PHPStan/Rules/Comparison/ConstantLooseComparisonRuleTest.php:192]: https://3v4l.org/W28WC

File: src/Type/IntegerType.php [hunk 4]
```diff
@@ -139,6 +140,18 @@ public function isScalar(): TrinaryLogic
 
 	public function looseCompare(Type $type, PhpVersion $phpVersion): BooleanType
```
Comment (staabm) [inline src/Type/IntegerType.php:141]: I think we have a new todo for PHPStan 3.x: adjust all `Type` methods which depend on `PhpVersion` and turn them into using `PhpVersions`

Comment (ondrejmirtes) [reply src/Type/IntegerType.php:141]: Yeah, this should be possible. It's possible to make PhpVersions from PhpVersion (without losing precision), but not vice-versa.

So places that only have PhpVersion could still call these methods by transforming it into PhpVersions first.

**Comment by ondrejmirtes** [comment]
Awesome, thank you!

**Pull request merged**